### PR TITLE
Publicize: Ensure that connections are only retrieved once per request.

### PR DESCRIPTION
**Pull request by pablinos**
A race condition where the transient, used to cache the connections information, gets set in a second request, while another is in flight, causes the `set_transient` call to fail.

As we relied on the transient, this meant that each time we tried to retrieve the current connection information, we would make a request to the API, and then fail to set the transient again.

This change solves that problem by storing the information in memory and also removing the transient if the call to `set_transient` fails.

There is an outstanding task where we need to work out why the Connections_Post_Field causes so many calls to `get_all_connections`.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
p1719382901926009-slack-C02JJ910CNL

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

This is a bit tricky as it's a race condition, but it seems to be quite possible with a local dev site

* Make sure you have at least one social connection on the site
* Create a test post, and line up your cursor with the left-hand side of the publish button (I know!)
* Click the button, and then quickly click the second Publish button that appears
* Without this change, the request will take quite a bit of time, and the post will not be shared to social media
* With this change, the requests should complete quickly, and the post will be shared as usual.

A video of this is in the linked Slack thread.

### CODE REVIEW

File: projects/packages/publicize/src/class-publicize.php [hunk 1]
```diff
@@ -138,7 +145,12 @@ public function disconnect( $service_name, $connection_id, $_blog_id = false, $_
 	 * @return true
 	 */
 	public function receive_updated_publicize_connections( $publicize_connections ) {
-		set_transient( self::JETPACK_SOCIAL_CONNECTIONS_TRANSIENT, $publicize_connections, 3600 * 4 );
+		$expiry = 3600 * 4;
+		if ( ! set_transient( self::JETPACK_SOCIAL_CONNECTIONS_TRANSIENT, $publicize_connections, $expiry ) ) {
+			delete_transient( self::JETPACK_SOCIAL_CONNECTIONS_TRANSIENT );
+			set_transient( self::JETPACK_SOCIAL_CONNECTIONS_TRANSIENT, $publicize_connections, $expiry );
+		}
+		$this->current_connections = $publicize_connections;
```
Comment (manzoorwanijk) [inline projects/packages/publicize/src/class-publicize.php]: An explanatory comment here can be helpful.

File: projects/packages/publicize/src/class-publicize.php [hunk 2]
```diff
@@ -138,7 +145,15 @@ public function disconnect( $service_name, $connection_id, $_blog_id = false, $_
 	 * @return true
 	 */
 	public function receive_updated_publicize_connections( $publicize_connections ) {
-		set_transient( self::JETPACK_SOCIAL_CONNECTIONS_TRANSIENT, $publicize_connections, 3600 * 4 );
+		$expiry = 3600 * 4;
+		if ( ! set_transient( self::JETPACK_SOCIAL_CONNECTIONS_TRANSIENT, $publicize_connections, $expiry ) ) {
+			// If the transient has beeen set in another request, the call to set_transient can fail. If so,
+			// we can delete the transient and try again.
+			delete_transient( self::JETPACK_SOCIAL_CONNECTIONS_TRANSIENT );
```
Comment (spsiddarthan) [inline projects/packages/publicize/src/class-publicize.php]: Nitpick, feel free to ignore: Should this be `$this->clear_connections_transient();`?

**Review by spsiddarthan — DISMISSED** [review]

**Comment by spsiddarthan** [comment]
Ah, I see some tests failing. Post-connections-field tests are a real mess to deal with.

**Review by manzoorwanijk — DISMISSED** [review]
Works well.

**Review by spsiddhu — APPROVED** [review]

**Review by spsiddarthan — APPROVED** [review]

**Pull request merged**